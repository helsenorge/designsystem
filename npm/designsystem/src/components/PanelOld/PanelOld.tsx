import React from 'react';

import classNames from 'classnames';

import { AnchorTarget, AnalyticsId } from '../../constants';
import { useExpand } from '../../hooks/useExpand';
import { useUuid } from '../../hooks/useUuid';
import { palette } from '../../theme/palette';
import { AriaLabelAttributes, getAriaLabelAttributes } from '../../utils/accessibility';
import Badge from '../Badge';
import Button, { ButtonProps, ButtonTags } from '../Button';
import Icon, { IconSize, SvgPathProps } from '../Icon';
import AlertSignFill from '../Icons/AlertSignFill';
import ArrowRight from '../Icons/ArrowRight';
import Calendar from '../Icons/Calendar';
import ChevronDown from '../Icons/ChevronDown';
import ChevronUp from '../Icons/ChevronUp';
import Pencil from '../Icons/Pencil';
import Watch from '../Icons/Watch';
import Title, { TitleTags } from '../Title';

import panelStyles from './styles.module.scss';

export enum PanelOldStatus {
  normal = 'normal',
  new = 'new',
  error = 'error',
  draft = 'draft',
}

export enum PanelOldVariant {
  fill = 'fill',
  white = 'white',
  stroke = 'stroke',
  line = 'line',
}

export enum PanelOldLayout {
  layout1 = 'layout1',
  layout2 = 'layout2',
  layout3a = 'layout3a',
  layout3b = 'layout3b',
  layout3c = 'layout3c',
}

export interface PanelOldProps {
  /** Adds custom classes to the element. */
  className?: string;
  /** Panel section A content */
  contentA?: React.ReactNode | string;
  /** Panel section B content */
  contentB?: React.ReactNode | string;
  /** Content for a container that renders above A and B regardless of layout */
  contentHeader?: React.ReactNode;
  /** Container acts as a button, clicking anywhere triggers a panel button click */
  containerAsButton?: boolean;
  /** Panel children */
  children?: React.ReactNode;
  /** Displays date with icon */
  date?: string;
  /** Expands or collapses the panel */
  expanded?: boolean;
  /** Whether the panel can be focused */
  focusable?: boolean;
  /** Icon displayed in title */
  icon?: React.ReactNode;
  /** Panel button text */
  buttonText?: string;
  /** Panel button close text */
  buttonTextClose?: string;
  /** HTML markup for panel button. Default: a */
  buttonHtmlMarkup?: ButtonTags;
  /** Callback when the panel button is clicked */
  buttonOnClick?: ButtonProps['onClick'];
  /** Panel button is aria-labelledby the text in the button itself + the element set in buttonAriaLabelledById. Default: auto-generated id for title (if title is set). */
  buttonAriaLabelledById?: string;
  /** Panel button aria label */
  buttonAriaLabel?: string;
  /** Layout of the panel */
  layout?: keyof typeof PanelOldLayout;
  /** Removes top border when variant is "line" */
  noTopBorder?: boolean;
  /** Called when the panel is opened/closed */
  onExpand?: (isExpanded: boolean) => void;
  /** Whether to render children when closed (in which case they are hidden with CSS). Default: false */
  renderChildrenWhenClosed?: boolean;
  /** Displays a status on the left side: default normal */
  status?: keyof typeof PanelOldStatus;
  /** Displayed on top of the panel with a status icon */
  statusMessage?: string;
  /** Sets the data-testid attribute for testing purposes */
  testId?: string;
  /** Displays time with icon */
  time?: string;
  /** Title of the panel */
  title?: string;
  /** Changes the underlying element of the title. Default: h2 */
  titleHtmlMarkup?: TitleTags;
  /** Changes the visual representation of the panel */
  variant?: keyof typeof PanelOldVariant;
  /** URL to details, renders as a button with anchor tag */
  url?: string;
  /** target used in the button: default is _self */
  target?: AnchorTarget;
}

export interface LayoutProps
  extends Pick<PanelOldProps, 'contentA' | 'contentB' | 'contentHeader' | 'icon' | 'layout' | 'status' | 'statusMessage'> {
  ctaContainer?: React.ReactNode;
  titleElement: React.ReactNode;
}

const StatusText: React.FC<{ status?: keyof typeof PanelOldStatus; statusMessage?: string }> = ({ status, statusMessage }) => {
  const statusIcon = (): { color: string; svgIcon: React.FC<SvgPathProps> } => {
    if (status === PanelOldStatus.error) {
      return { color: palette.cherry500, svgIcon: AlertSignFill };
    }

    return { color: palette.black, svgIcon: Pencil };
  };

  const statusMessageClass = classNames(panelStyles['status-message'], {
    [panelStyles['status-message--new']]: status === PanelOldStatus.new,
  });

  if ((status === PanelOldStatus.error || status === PanelOldStatus.draft) && statusMessage) {
    return (
      <div className={statusMessageClass} data-testid="display-status">
        {<Icon {...statusIcon()} size={IconSize.XSmall} />} <span>{statusMessage}</span>
      </div>
    );
  }

  return null;
};

const PreContainer: React.FC<{ children?: React.ReactNode }> = ({ children }) => {
  if (typeof children === 'undefined') return null;
  return <div className={panelStyles['header-container']}>{children}</div>;
};

const DateTime: React.FC<{ date?: string; time?: string }> = ({ date, time }) => {
  if (date || time) {
    return (
      <div className={panelStyles['datetime-container']} data-testid="datetime">
        {date && (
          <div className={panelStyles['datetime-container__icon']}>
            <Icon svgIcon={Calendar} size={IconSize.XSmall} />
            <span>{date}</span>
          </div>
        )}
        {time && (
          <div className={panelStyles['datetime-container__icon']}>
            <Icon svgIcon={Watch} size={IconSize.XSmall} />
            <span>{time}</span>
          </div>
        )}
      </div>
    );
  }

  return null;
};

const PanelLayout1: React.FC<LayoutProps> = ({
  contentA,
  contentB,
  contentHeader,
  ctaContainer,
  icon,
  status,
  statusMessage,
  titleElement,
}) => {
  const panelLayoutClasses = classNames(panelStyles['panel__layout-1'], {
    [panelStyles['panel__layout-1--with-icon']]: icon,
  });
  const iconClasses = classNames(panelStyles.panel__icon, panelStyles['panel__icon--layout-1'], {
    [panelStyles['panel__icon--no-content']]: !contentA && !contentB,
  });

  return (
    <div className={panelLayoutClasses}>
      <StatusText status={status} statusMessage={statusMessage} />
      <PreContainer>{contentHeader}</PreContainer>
      {icon && <div className={iconClasses}>{icon}</div>}
      <div className={panelStyles['panel__layout-1__content-a']}>
        {titleElement}
        {contentA}
      </div>
      {contentB && <div>{contentB}</div>}
      {ctaContainer}
    </div>
  );
};

const PanelLayout2: React.FC<LayoutProps> = ({
  contentA,
  contentB,
  contentHeader,
  ctaContainer,
  icon,
  status,
  statusMessage,
  titleElement,
}) => {
  const panelLayoutClasses = classNames(panelStyles['panel__layout-2'], {
    [panelStyles['panel__layout-2--with-icon']]: icon,
  });
  const iconClasses = classNames(panelStyles.panel__icon, panelStyles['panel__icon--layout-2'], {
    [panelStyles['panel__icon--no-content']]: !contentA && !contentB,
  });
  const lastColumnClass = panelStyles['panel__layout-2__last-column'];

  return (
    <div className={panelLayoutClasses}>
      <StatusText status={status} statusMessage={statusMessage} />
      <PreContainer>{contentHeader}</PreContainer>
      {icon && <div className={iconClasses}>{icon}</div>}
      <div className={panelStyles['panel__layout-2__content-a']}>
        {titleElement}
        {contentA}
      </div>
      {contentB && <div className={lastColumnClass}>{contentB}</div>}
      {ctaContainer && <div className={lastColumnClass}>{ctaContainer}</div>}
    </div>
  );
};

const PanelLayout3: React.FC<LayoutProps> = ({
  contentA,
  contentB,
  contentHeader,
  ctaContainer,
  icon,
  layout,
  status,
  statusMessage,
  titleElement,
}) => {
  const layoutClasses = classNames(panelStyles['panel__layout-3'], {
    [panelStyles['panel__layout-3--with-icon']]: icon,
    [panelStyles['panel__layout-3--a']]: layout === PanelOldLayout.layout3a,
    [panelStyles['panel__layout-3--b']]: layout === PanelOldLayout.layout3b,
    [panelStyles['panel__layout-3--c']]: layout === PanelOldLayout.layout3c,
  });
  const iconClasses = classNames(panelStyles.panel__icon, panelStyles['panel__icon--layout-3'], {
    [panelStyles['panel__icon--no-content']]: !contentA && !contentB,
  });

  return (
    <div className={layoutClasses}>
      <StatusText status={status} statusMessage={statusMessage} />
      <PreContainer>{contentHeader}</PreContainer>
      {icon && <div className={iconClasses}>{icon}</div>}
      <div>
        {titleElement}
        {contentA}
      </div>
      <div className={panelStyles['panel__layout-3__last-column']}>
        {contentB && <div className={panelStyles['panel__layout-3__last-column__content-b']}>{contentB}</div>}
        {ctaContainer}
      </div>
    </div>
  );
};

const PanelOld = React.forwardRef(function PanelForwardedRef(props: PanelOldProps, ref: React.ForwardedRef<HTMLHeadingElement>) {
  const {
    buttonAriaLabel,
    buttonAriaLabelledById,
    buttonText = 'Se detaljer',
    buttonTextClose = 'Skjul detaljer',
    buttonHtmlMarkup = 'a',
    buttonOnClick,
    children,
    className,
    containerAsButton = false,
    contentA,
    contentB,
    contentHeader,
    date,
    expanded = false,
    focusable = false,
    icon,
    layout = PanelOldLayout.layout2,
    noTopBorder,
    onExpand,
    renderChildrenWhenClosed = false,
    status = PanelOldStatus.normal,
    statusMessage,
    target = '_self',
    testId,
    time,
    title,
    titleHtmlMarkup = 'h2',
    url,
    variant = PanelOldVariant.fill,
  } = props;

  const [isExpanded, setIsExpanded] = useExpand(expanded, onExpand);
  const titleId = useUuid();
  const buttonTextId = useUuid();
  const hasBadge = statusMessage && status === PanelOldStatus.new;
  const noContentB = typeof contentB === 'undefined';
  const layout1 = layout === 'layout1' || noContentB;
  const layout2 = !noContentB && layout === 'layout2';
  const layout3 = !noContentB && (layout === 'layout3a' || layout === 'layout3b' || layout === 'layout3c');
  const panelWrapperClasses = classNames(panelStyles['panel-wrapper'], className);

  const panelClasses = classNames(panelStyles.panel, {
    [panelStyles['panel--fill']]: variant === PanelOldVariant.fill,
    [panelStyles['panel--stroke']]: variant === PanelOldVariant.stroke,
    [panelStyles['panel--white']]: variant === PanelOldVariant.white,
    [panelStyles['panel--line']]: variant === PanelOldVariant.line,
    [panelStyles['panel--no-top-border']]: variant === PanelOldVariant.line && noTopBorder,
    [panelStyles['panel--selected']]: isExpanded,
    [panelStyles['panel--new']]: status === PanelOldStatus.new,
    [panelStyles['panel--draft']]: status === PanelOldStatus.draft,
    [panelStyles['panel--error']]: status === PanelOldStatus.error,
    [panelStyles['panel--status']]: status && status !== PanelOldStatus.normal,
    [panelStyles['panel--with-icon']]: icon,
    [panelStyles['panel--button']]: containerAsButton,
    [panelStyles['panel--clickable']]: children || url || onExpand || buttonOnClick || containerAsButton,
  });

  const renderCTAContainer = () => {
    const hasButton = children || url || buttonOnClick;
    const btnContainerClasses = classNames(panelStyles['panel__btn-container'], {
      [panelStyles['panel__btn-container--no-content-b']]: noContentB,
      [panelStyles['panel__btn-container--no-button']]: !hasButton,
    });

    return (
      (hasButton || date || time) && (
        <div className={btnContainerClasses}>
          {<DateTime date={date} time={time} />}
          {hasButton && <div className={panelStyles['panel__details-btn']}>{renderDetailsButton()}</div>}
        </div>
      )
    );
  };

  const renderDetailsButton = (): React.ReactNode => {
    const ariaLabelAttributes = getAriaLabelAttributes({
      label: buttonAriaLabel,
      id: (buttonAriaLabelledById && `${buttonTextId} ${buttonAriaLabelledById}`) || (title && titleId && `${buttonTextId} ${titleId}`),
      prefer: 'label',
    });

    const commonProps: Partial<ButtonProps> & AriaLabelAttributes = {
      onClick: buttonOnClick ? buttonOnClick : (): void => setIsExpanded(!isExpanded),
      className: containerAsButton ? panelStyles['panel__expand'] : undefined,
      variant: 'borderless',
      ellipsis: true,
      ...ariaLabelAttributes,
    };

    if (children) {
      return (
        <Button testId="expand" aria-expanded={isExpanded} {...commonProps}>
          <span id={buttonTextId}>{isExpanded ? buttonTextClose : buttonText}</span>
          <Icon svgIcon={isExpanded ? ChevronUp : ChevronDown} />
        </Button>
      );
    }

    return (
      <Button testId="url" htmlMarkup={buttonHtmlMarkup} href={url} target={target} {...commonProps}>
        <span id={buttonTextId}>{buttonText}</span>
        <Icon svgIcon={ArrowRight} />
      </Button>
    );
  };

  const renderContent = (): React.ReactNode | null => {
    if (!children || (!renderChildrenWhenClosed && !isExpanded)) {
      return null;
    }

    const panelDetailsClasses = classNames(panelStyles['panel-details'], {
      [panelStyles['panel-details--open']]: isExpanded,
      [panelStyles['panel-details--line']]: variant === PanelOldVariant.line,
      [panelStyles['panel-details--white']]: variant === PanelOldVariant.white,
    });

    return (
      <div className={panelDetailsClasses} data-testid="panel-details">
        <div>{children}</div>
      </div>
    );
  };

  const renderTitle = () => {
    const titleContainerClasses = classNames(panelStyles['title-container'], {
      [panelStyles['title-container--no-content-a']]: !contentA,
    });
    const titleClasses = classNames(panelStyles['title-container__title'], {
      [panelStyles['title-container__title--badge']]: hasBadge,
    });

    return (
      title && (
        <div className={titleContainerClasses}>
          <Title appearance="title3" htmlMarkup={titleHtmlMarkup} id={titleId} className={titleClasses}>
            {title}
          </Title>
          {hasBadge && (
            <div className={panelStyles.panel__badge}>
              <Badge color="blueberry" testId="badge-status">
                {statusMessage}
              </Badge>
            </div>
          )}
        </div>
      )
    );
  };

  const layoutProps: LayoutProps = {
    contentA: contentA,
    contentB: contentB,
    contentHeader: contentHeader,
    ctaContainer: renderCTAContainer(),
    icon: icon,
    status: status,
    statusMessage: statusMessage,
    titleElement: renderTitle(),
  };

  return (
    <div
      // eslint-disable-next-line no-constant-condition
      tabIndex={focusable ? -1 : undefined}
      ref={ref}
      data-testid={testId}
      className={panelWrapperClasses}
      data-analyticsid={AnalyticsId.Panel}
    >
      <div className={panelClasses}>
        {layout1 && <PanelLayout1 {...layoutProps} />}
        {layout2 && <PanelLayout2 {...layoutProps} />}
        {layout3 && <PanelLayout3 {...layoutProps} layout={layout} />}
      </div>
      {renderContent()}
    </div>
  );
});

export default PanelOld;
