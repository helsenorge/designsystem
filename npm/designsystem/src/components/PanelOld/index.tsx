import PanelOld from './PanelOld';
export * from './PanelOld';
export default PanelOld;
