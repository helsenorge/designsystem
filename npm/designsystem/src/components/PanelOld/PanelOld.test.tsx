import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import PanelOld from './PanelOld';
import * as uuidUtils from '../../utils/uuid';
import Avatar from '../Avatar';
import { Icon, IconSize } from '../Icon';
import Calendar from '../Icons/Calendar';

describe('Gitt at PanelOld skal vises', (): void => {
  describe('Når testId-prop er satt', (): void => {
    test('Så kan komponenten finnes ved hjelp av testId', (): void => {
      render(<PanelOld testId="bare-tester" />);
      const component = screen.getByTestId('bare-tester');
      expect(component).toBeVisible();
      expect(component).toHaveAttribute('data-analyticsid', 'panel');
    });
  });

  describe('Når panelet har children', (): void => {
    describe('Når panelet har title', (): void => {
      test('Så er tittelen synlig', (): void => {
        render(
          <PanelOld title="Medisinsk fødselsregister" buttonText="Vis detaljer">
            <p>{'Mer tekst'}</p>
          </PanelOld>
        );
        const title = screen.getByRole('heading', { name: 'Medisinsk fødselsregister' });
        expect(title).toBeVisible();
      });
      test('Så har knappen som brukes for å ekspandere et unikt navn', (): void => {
        vi.spyOn(uuidUtils, 'uuid').mockReturnValueOnce('a').mockReturnValueOnce('b');

        render(
          <PanelOld title="Medisinsk fødselsregister" buttonText="Vis detaljer">
            <p>{'Mer tekst'}</p>
          </PanelOld>
        );
        const button = screen.getByRole('button', { name: 'Vis detaljer Medisinsk fødselsregister' });
        expect(button).toBeVisible();
      });
    });
    describe('Når panelet ikke har title', (): void => {
      test('Så har knappen som brukes for å ekspandere riktig navn', (): void => {
        vi.spyOn(uuidUtils, 'uuid').mockReturnValueOnce('a').mockReturnValueOnce('b');

        render(
          <PanelOld buttonText="Vis detaljer">
            <p>{'Mer tekst'}</p>
          </PanelOld>
        );
        const button = screen.getByRole('button', { name: 'Vis detaljer' });
        expect(button).toBeVisible();
      });
    });
  });

  describe('Når panelet har url', (): void => {
    describe('Når panelet har title', (): void => {
      test('Så har linken et unikt navn', (): void => {
        vi.spyOn(uuidUtils, 'uuid').mockReturnValueOnce('a').mockReturnValueOnce('b');

        render(<PanelOld title="Medisinsk fødselsregister" buttonText="Åpne" url={'https://www.helsenorge.no'} />);

        const link = screen.getByRole('link', { name: 'Åpne Medisinsk fødselsregister' });
        expect(link).toBeVisible();
      });
    });
  });

  describe('Når buttonAriaLabelledById-prop er satt', (): void => {
    test('Så har knappen som brukes for å ekspandere riktig navn', (): void => {
      render(
        <PanelOld buttonText="Vis detaljer" buttonAriaLabelledById="egen-id" contentA={<h2 id="egen-id">{'Egendefinert tittel'}</h2>}>
          <p>{'Mer tekst'}</p>
        </PanelOld>
      );
      const button = screen.getByRole('button', { name: 'Vis detaljer Egendefinert tittel' });
      expect(button).toBeVisible();
    });
  });

  describe('Når buttonAriaLabel-prop er satt', (): void => {
    test('Så har knappen som brukes for å ekspandere riktig navn', (): void => {
      render(
        <PanelOld buttonText="Vis detaljer" buttonAriaLabel="Helt egen tekst på knappen">
          <p>{'Mer tekst'}</p>
        </PanelOld>
      );
      const button = screen.getByRole('button', { name: 'Helt egen tekst på knappen' });
      expect(button).toBeVisible();
    });
  });

  describe('Når title og  buttonAriaLabel-prop er satt', (): void => {
    test('Så har knappen som brukes for å ekspandere riktig navn', (): void => {
      render(
        <PanelOld title="Medisinsk fødselsregister" buttonText="Vis detaljer" buttonAriaLabel="Helt egen tekst på knappen">
          <p>{'Mer tekst'}</p>
        </PanelOld>
      );
      const button = screen.getByRole('button', { name: 'Helt egen tekst på knappen' });
      expect(button).toBeVisible();
    });
  });

  describe('Når titleHtmlMarkup-prop er satt', (): void => {
    test('Så er tittel satt til ønsket overskriftsnivå', (): void => {
      render(<PanelOld title="Tittel" titleHtmlMarkup="h3" />);
      const title = screen.getByRole('heading', { name: 'Tittel' });
      expect(title).toBeVisible();
      expect(title.tagName).toEqual('H3');
    });
  });

  describe('Når icon er satt', (): void => {
    test('Så skal komponenten vise ikon på venstre side', (): void => {
      render(<PanelOld testId="bare-tester" icon={<Icon testId="icon-testid" svgIcon={Calendar} size={IconSize.XSmall} />} />);
      const icon = screen.getByTestId('icon-testid');
      expect(icon).toBeVisible();
      // eslint-disable-next-line testing-library/no-node-access
      expect(icon.parentElement).toHaveClass('panel__icon');
    });
  });

  describe('Når Avatar er satt', (): void => {
    test('Så skal komponenten vise Avatar', (): void => {
      render(<PanelOld icon={<Avatar testId="avatar-test">{'Anne Brun'}</Avatar>} />);
      const avatar = screen.getByTestId('avatar-test');
      expect(avatar).toBeVisible();
      // eslint-disable-next-line testing-library/no-node-access
      expect(avatar.parentElement).toHaveClass('panel__icon');
    });
  });

  describe('Gitt at status ikke er satt', (): void => {
    test('Så skal PanelOldvises som default normal', (): void => {
      render(<PanelOld testId="panel-test-status" title="Dette er en tittel" />);
      const panel = screen.getByTestId('panel-test-status');
      // eslint-disable-next-line testing-library/no-node-access
      expect(panel.firstChild).toHaveClass('panel panel--fill');
    });
  });

  describe('Gitt at status er new og statusMessage er satt', (): void => {
    test('Så vises panel i new visning og Badge vises med statusMessage', (): void => {
      render(<PanelOld testId="panel-test-status" status="new" statusMessage="1 ny" title="Dette er en tittel" />);
      const panel = screen.getByTestId('panel-test-status');
      const title = screen.getByText('Dette er en tittel');
      const badge = screen.getByTestId('badge-status');

      // eslint-disable-next-line testing-library/no-node-access
      expect(panel.firstChild).toHaveClass('panel panel--fill panel--new panel--status');
      expect(title).toHaveClass('title title--title3 title-container__title title-container__title--badge');
      // eslint-disable-next-line testing-library/no-node-access
      expect(badge.parentElement).toHaveClass('panel__badge');
      expect(badge.innerHTML).toEqual('1 ny');
    });
  });

  describe('Gitt at status er draft og statusMessage er satt', (): void => {
    test('Så vises panel i draft visning og statusMessage vises med ikon', (): void => {
      render(<PanelOld testId="panel-test-status" status="draft" statusMessage="Dette en beskjed" title="Dette er en tittel" />);
      const panel = screen.getByTestId('panel-test-status');
      // eslint-disable-next-line testing-library/no-node-access
      expect(panel.firstChild).toHaveClass('panel panel--fill panel--draft panel--status');
      const status = screen.getByTestId('display-status');
      // eslint-disable-next-line testing-library/no-node-access
      expect(status.children[1].innerHTML).toEqual('Dette en beskjed');

      //expect draft icon
      expect(status).toMatchSnapshot();
    });
  });

  describe('Gitt at status er error og statusMessage er satt', (): void => {
    test('Så vises panel i error visning og statusMessage vises med ikon', (): void => {
      render(<PanelOld testId="panel-test-status" status="error" statusMessage="Dette en error beskjed" title="Dette er en tittel" />);
      const panel = screen.getByTestId('panel-test-status');
      // eslint-disable-next-line testing-library/no-node-access
      expect(panel.firstChild).toHaveClass('panel panel--fill panel--error panel--status');
      const status = screen.getByTestId('display-status');
      // eslint-disable-next-line testing-library/no-node-access
      expect(status.children[1].innerHTML).toEqual('Dette en error beskjed');

      //expect error icon
      expect(status).toMatchSnapshot();
    });
  });

  describe('Gitt at panelet har default variant', (): void => {
    test('Så skal panelet vise en default variant', (): void => {
      render(<PanelOld testId="panel-test-status" title="Dette er en tittel" />);
      const panel = screen.getByTestId('panel-test-status');
      // eslint-disable-next-line testing-library/no-node-access
      expect(panel.firstChild).toHaveClass('panel panel--fill');
    });
  });

  describe('Gitt at panelet har variant line', (): void => {
    test('Så skal panelet vise variant line', (): void => {
      render(<PanelOld variant="line" testId="panel-test-status" title="Dette er en tittel" />);
      const panel = screen.getByTestId('panel-test-status');
      // eslint-disable-next-line testing-library/no-node-access
      expect(panel.firstChild).toHaveClass('panel panel--line');
    });
  });

  describe('Gitt at panelet har variant stroke', (): void => {
    test('Så skal panelet vise variant stroke', (): void => {
      render(<PanelOld variant="stroke" testId="panel-test-status" title="Dette er en tittel" />);
      const panel = screen.getByTestId('panel-test-status');
      // eslint-disable-next-line testing-library/no-node-access
      expect(panel.firstChild).toHaveClass('panel panel--stroke');
    });
  });

  describe('Gitt at panelet har variant white', (): void => {
    test('Så skal panelet vise variant white', (): void => {
      render(<PanelOld variant="white" testId="panel-test-status" title="Dette er en tittel" />);
      const panel = screen.getByTestId('panel-test-status');
      // eslint-disable-next-line testing-library/no-node-access
      expect(panel.firstChild).toHaveClass('panel panel--white');
    });
  });

  describe('Gitt at panelet har en url knapp', (): void => {
    test('Så skal knappen rendres som anchor-tag', (): void => {
      render(<PanelOld url="/#?test" title="Dette er en tittel" />);
      const btnDetails = screen.getByTestId('url');
      expect(btnDetails).toHaveAttribute('href', '/#?test');
      expect(btnDetails).toHaveAttribute('target', '_self');
    });
  });

  describe('Gitt at panelet har en expand knapp, og gitt at knappen trykkes', (): void => {
    test('Så vises detalje-området', async (): Promise<void> => {
      render(
        <PanelOld title="Dette er en tittel" renderChildrenWhenClosed>
          <div>{'Details'}</div>
        </PanelOld>
      );

      const btnDetails = screen.getByTestId('expand');

      expect(btnDetails).toBeVisible();

      const panelDetails = screen.getByTestId('panel-details');

      await userEvent.click(btnDetails);
      expect(screen.getByText('Details')).toBeVisible();
      expect(panelDetails).toHaveClass('panel-details--open');

      await userEvent.click(btnDetails);
      expect(panelDetails).not.toHaveClass('panel-details--open');
    });
  });

  describe('Gitt at panelet skal vises som ekspandert', (): void => {
    test('Så vises innholdet ekspandert fra starten av', (): void => {
      render(
        <PanelOld title={'Dette er en tittel'} expanded>
          <h1>{'Innhold i panel'}</h1>
        </PanelOld>
      );

      const button = screen.getByTestId('expand');
      expect(button).toHaveAttribute('aria-expanded', 'true');

      const content = screen.getByRole('heading', { name: 'Innhold i panel' });
      expect(content).toBeInTheDocument();
    });
    test('Så er det bare én knapp for skjul detaljer', (): void => {
      render(
        <PanelOld title={'Dette er en tittel'} expanded>
          <h1>{'Innhold i panel'}</h1>
        </PanelOld>
      );

      const button = screen.getByTestId('expand');
      expect(button).toBeVisible();
    });
  });

  describe('Gitt at panelet har onExpand-callback', (): void => {
    describe('Når man klikker på panelet to ganger', (): void => {
      test('Så kalles callback først med true og så med false', async (): Promise<void> => {
        const handleExpand = vi.fn();
        render(
          <PanelOld title={'Dette er en tittel'} onExpand={handleExpand}>
            <h1>{'Innhold i panel'}</h1>
          </PanelOld>
        );

        const button = screen.getAllByTestId('expand')[0];

        await userEvent.click(button);
        await userEvent.click(button);

        expect(handleExpand).toHaveBeenCalledTimes(2);
        expect(handleExpand).toHaveBeenNthCalledWith(1, true);
        expect(handleExpand).toHaveBeenNthCalledWith(2, false);
      });
    });
  });
  describe('Gitt at panelet har onExpand-callback og expanded satt til true', (): void => {
    describe('Når man klikker på panelet én gang', (): void => {
      test('Så kalles callback først med true og så med false', async (): Promise<void> => {
        const handleExpand = vi.fn();
        render(
          <PanelOld title={'Dette er en tittel'} onExpand={handleExpand} expanded>
            <h1>{'Innhold i panel'}</h1>
          </PanelOld>
        );

        const button = screen.getAllByTestId('expand')[0];

        await userEvent.click(button);

        expect(handleExpand).toHaveBeenCalledTimes(2);
        expect(handleExpand).toHaveBeenNthCalledWith(1, true);
        expect(handleExpand).toHaveBeenNthCalledWith(2, false);
      });
    });
  });

  describe('Gitt at containerAsButton er true', (): void => {
    describe('Når man klikker på knappen for å vise detaljer', (): void => {
      test('Så vises detaljer-området', async (): Promise<void> => {
        render(
          <PanelOld testId="panel-test" containerAsButton title="Dette er en tittel" renderChildrenWhenClosed>
            <div>{'Details'}</div>
          </PanelOld>
        );

        const button = screen.getByTestId('expand');

        expect(button).toBeVisible();
        expect(button).toHaveAttribute('aria-expanded', 'false');

        await userEvent.click(button);

        expect(screen.getByText('Details')).toBeVisible();
      });
    });
  });

  describe('Gitt at panelet skal vise tid og dato', (): void => {
    test('Så vises tid og dato', (): void => {
      render(
        <PanelOld testId="panel-test" date="dato" time="tid" title="Dette er en tittel" contentA={'Innhold A'} contentB={'Innhold B'}>
          <div>{'Details'}</div>
        </PanelOld>
      );

      expect(screen.getByText('dato')).toBeVisible();
      expect(screen.getByText('tid')).toBeVisible();
    });
  });

  describe('Gitt at panelet skal vise lenke som knapp med onClick', () => {
    test('Så kalles onClick-handler når man klikker på knappen', async () => {
      const onClickMock = vi.fn();
      render(<PanelOld buttonText="Lenke til mer" buttonOnClick={onClickMock} buttonHtmlMarkup="button" />);

      const link = screen.getByRole('button', { name: 'Lenke til mer' });
      await userEvent.click(link);
      expect(onClickMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Gitt at panelet skal vise lenke som a med onClick', () => {
    test('Så kalles onClick-handler når man klikker på lenken', async () => {
      const onClickMock = vi.fn();
      render(<PanelOld buttonText="Lenke til mer" buttonOnClick={onClickMock} buttonHtmlMarkup="a" url="#" />);

      const link = screen.getByRole('link', { name: 'Lenke til mer' });

      expect(link).toHaveAttribute('href', '#');

      await userEvent.click(link);
      expect(onClickMock).toHaveBeenCalledTimes(1);
    });
  });
  describe('Når renderChildrenWhenClosed ikke er satt', (): void => {
    test('Så er children ikke rendret', (): void => {
      render(
        <PanelOld>
          <span data-testid="test">{'Test'}</span>
        </PanelOld>
      );

      const child = screen.queryByTestId('test');
      expect(child).not.toBeInTheDocument();
    });
  });
  describe('Når renderChildrenWhenClosed er true', (): void => {
    test('Så er children rendret', (): void => {
      render(
        <PanelOld renderChildrenWhenClosed={true}>
          <span data-testid="test">{'Test'}</span>
        </PanelOld>
      );

      const child = screen.getByTestId('test');
      expect(child).toBeInTheDocument();
    });
  });
  describe('Når renderChildrenWhenClosed er false', (): void => {
    test('Så er children ikke rendret', (): void => {
      render(
        <PanelOld renderChildrenWhenClosed={false}>
          <span data-testid="test">{'Test'}</span>
        </PanelOld>
      );

      const child = screen.queryByTestId('test');
      expect(child).not.toBeInTheDocument();
    });
  });
});
