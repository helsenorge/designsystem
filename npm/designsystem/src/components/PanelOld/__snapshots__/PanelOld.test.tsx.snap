// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gitt at PanelOld skal vises > Gitt at status er draft og statusMessage er satt > SÃ¥ vises panel i draft visning og statusMessage vises med ikon 1`] = `
<div
  class="status-message"
  data-testid="display-status"
>
  <svg
    aria-hidden="true"
    class="hnds-style-icon"
    color="#000000"
    data-analyticsid="icon"
    fill="#000000"
    focusable="false"
    height="38"
    role="presentation"
    style="min-width: 38px; min-height: 38px;"
    viewBox="0 0 48 48"
    width="38"
  >
    <path
      d="M38.188 15.243l-.754.754-5.702-5.703.753-.755c1.522-1.523 4.178-1.522 5.703 0a4.04 4.04 0 010 5.704zm-3.08 3.08l-5.702-5.704 1.61-1.61 5.703 5.703-1.61 1.61zM19.796 33.633l-2.49-2.49 14.702-14.493 2.387 2.385-14.599 14.598zm-6.856 3.28l-2.117-2.118 2.313-5.32 5.118 5.118-5.314 2.32zm-3.75 1.638l1.2-2.758 1.554 1.555-2.753 1.203zm19.501-25.218l2.602 2.602L16.592 30.43l-2.499-2.499 14.598-14.597zM39.438 8.29V8.29a5.768 5.768 0 00-4.102-1.699 5.765 5.765 0 00-4.102 1.699L12.216 27.306a.895.895 0 00-.184.274L6.673 39.91a.884.884 0 001.165 1.163l12.312-5.377a.888.888 0 00.271-.186l19.017-19.017a5.808 5.808 0 000-8.203z"
    />
  </svg>
   
  <span>
    Dette en beskjed
  </span>
</div>
`;

exports[`Gitt at PanelOld skal vises > Gitt at status er error og statusMessage er satt > SÃ¥ vises panel i error visning og statusMessage vises med ikon 1`] = `
<div
  class="status-message"
  data-testid="display-status"
>
  <svg
    aria-hidden="true"
    class="hnds-style-icon"
    color="#C83521"
    data-analyticsid="icon"
    fill="#C83521"
    focusable="false"
    height="38"
    role="presentation"
    style="min-width: 38px; min-height: 38px;"
    viewBox="0 0 48 48"
    width="38"
  >
    <path
      d="M24.896 21.452v7.213a.891.891 0 01-1.782 0v-7.213a.89.89 0 111.782 0zm.5 11.577a1.39 1.39 0 11-2.782-.001 1.39 1.39 0 012.782.001zM6.613 39.725h34.774L24.11 8.275 6.613 39.725z"
    />
  </svg>
   
  <span>
    Dette en error beskjed
  </span>
</div>
`;
